aes-kw/src/kek.rs

```diff
@@ -254,4 +281,164 @@ where
+    pub fn wrap_with_padding_vec(&self, data: &[u8]) -> Result<Vec<u8>> {
+        let n = (data.len() + SEMIBLOCK_SIZE - 1) / SEMIBLOCK_SIZE;
+        let mut out = vec![0u8; n * SEMIBLOCK_SIZE + IV_LEN];
+        self.wrap_with_padding(data, &mut out)?;
+        Ok(out)
+    }
+
+    /// AES Key Wrap with Padding, as defined in RFC 5649.
+    ///
+    /// The `out` buffer will be overwritten, and must be exactly [`IV_LEN`]
+    /// bytes (i.e. 8 bytes) shorter than the length of `data`.
+    /// This method returns a slice of `out`, truncated to the appropriate
+    /// length by removing the padding.
+    pub fn unwrap_with_padding<'a>(&self, data: &[u8], out: &'a mut [u8]) -> Result<&'a [u8]> {
+        if data.len() % SEMIBLOCK_SIZE != 0 {
+            return Err(Error::InvalidDataSize);
+        }
+
+        // 0) Prepare inputs
+
+        let n = (data.len() / SEMIBLOCK_SIZE)
+            .checked_sub(1)
+            .ok_or(Error::InvalidDataSize)?;
+
+        if out.len() != n * SEMIBLOCK_SIZE {
+            return Err(Error::InvalidOutputSize {
+                expected: n * SEMIBLOCK_SIZE,
+            });
+        }
+
+        // 1) Key unwrapping
+
+        // 1.1) Initialize variables
+
+        let mut block = GenericArray::<u8, Aes::BlockSize>::default();
+
+        // If n is 1, the plaintext is encrypted as a single AES block
+        if n == 1 {
+            block.copy_from_slice(data);
+
+            self.cipher.decrypt_block(&mut block);
+            out.copy_from_slice(&block[IV_LEN..]);
+        } else {
+            block[..IV_LEN].copy_from_slice(&data[..IV_LEN]);
+
+            //   for i = 1 to n: R[i] = C[i]
+            out.copy_from_slice(&data[IV_LEN..]);
+
+            // 1.2) Calculate intermediate values
+
+            self.w_inverse(n, &mut block, out);
+        }
+
+        // 2) AIV verification
+
+        // Checks as defined in RFC5649 § 3
+
+        if block[..IV_LEN / 2] != IV_PREFIX {
+            return Err(Error::IntegrityCheckFailed);
+        }
+
+        let mli = u32::from_be_bytes(block[IV_LEN / 2..IV_LEN].try_into().unwrap()) as usize;
+        if !(8 * (n - 1) < mli && mli <= 8 * n) {
+            return Err(Error::IntegrityCheckFailed);
+        }
+
+        let b = 8 * n - mli;
+        if !out.iter().rev().take(b).all(|&x| x == 0) {
+            return Err(Error::IntegrityCheckFailed);
+        }
+
+        // 3) Output the results
+
+        Ok(&out[..mli])
+    }
+
+    /// Computes [`Self::unwrap`], allocating a [`Vec`] for the return value.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    pub fn unwrap_with_padding_vec(&self, data: &[u8]) -> Result<Vec<u8>> {
+        let out_len = data
+            .len()
+            .checked_sub(IV_LEN)
+            .ok_or(Error::InvalidDataSize)?;
+
+        let mut out = vec![0u8; out_len];
+        let out_slice = self.unwrap_with_padding(data, &mut out)?;
+        Ok(out_slice.to_vec())
```

Comment:
What's the performance of `to_vec` here? Are we allocating a new Vec from the `out_slice`? Or is Rust smart enough to reuse the part of `out` that's actually in `out_slice`?

Reply:
I'd suggest reusing the vector and truncating the output here. Something like...

```rust
        let mut out = vec![0u8; out_len];
        let out_len = self.unwrap_with_padding(data, &mut out)?.len();
        out.truncate(out_len);
        Ok(out)
```